### ISSUE TITLE
Add blosc2_decompress_safe()

### ISSUE THREAD

**Issue by FrancescAlted** [Blosc/c-blosc2#71]
Following the discussion in #40, it could be handy to have a new call for safer decompression that might have this signature:

```
int blosc2_decompress_safe(blosc2_context* context, const void* src, size_t *srcsize, void* dest, size_t *destsize);
```

On input, `srcsize` and `destsize` will have maximum number of bytes for reading (`srcsize`) or writing (`destsize`), whereas on output, these will contain the actual number of bytes read (`srcsize`) and written (`destsize`).

Essentially, the implementation would be something like:

```
int blosc2_decompress_safe(blosc2_context* context, const void* src, size_t* srcsize, void* dest, size_t* destsize);
  int result;
  size_t nbytes, cbytes, blocksize;

  blosc_cbuffer_sizes(src, &nbytes,  &cbytes, &blocksize);
  if (srcsize < cbytes) {
    *srcbytes = BLOSC_EXTENDED_HEADER_LENGTH;
    *destsize = 0;
    return -11;  // means that srcsize is too short
   }
  if (destsize < nbytes) {
    *srcbytes = BLOSC_EXTENDED_HEADER_LENGTH;
    *destsize = 0;
    return -12;  // means that destsize is too short
   }

  result = blosc2_decompress_ctx(context, src, dest, destsize);
  if (result > 0) {
    *srcsize = cbytes;
    *destsize = nbytes;
  }
  return result;
}
```

**Comment by FrancescAlted** [comment]
Oops, I was writing too fast I guess.  Corrected now.  That makes sense now?

**Comment by FrancescAlted** [comment]
I think it would be better to discuss over an on going PR.  Would you mind creating one?

**Comment by jinfeihan57** [comment]
I want to help to solve this issue. Is that OK ? 
But the solution may be the same as the one mentioned above.

**Comment by FrancescAlted** [comment]
@nmoinvaz any input on this?

**Comment by FrancescAlted** [comment]
I am thinking that perhaps we could add kind of a fingerprint at the end of either the header or the chunk so that we can verify that the contents of the header/chunk have not been tampered.  This supposes breaking the existing Blosc2 format, but are still in beta, so perhaps worth considering.

**Comment by jinfeihan57** [comment]
what kand of figerprint?
If you need to 100% verify the correctness of the input data, you must decode it and use CRC or Hash verification. 
If so, it is implemented in the decoding function. There is no way to determine in advance that the input data is 100% correct. Unless it is decoded and CRC or hash check is used.

**Comment by nmoinvaz** [comment]
Yes I am currently working on this. It is not enough to do a simple function like above because we have to check each time we seek to a new position in srcbuf to see if it is safe. If you want to assign this GH issue to me feel free.
